lib/kafka/round_robin_assignment_strategy.rb

```diff
@@ -24,20 +24,28 @@ def assign(members:, topics:)
         group_assignment[member_id] = Protocol::MemberAssignment.new
       end
 
-      topics.each do |topic|
+      topic_partitions = topics.flat_map do |topic|
         begin
           partitions = @cluster.partitions_for(topic).map(&:partition_id)
         rescue UnknownTopicOrPartition
           raise UnknownTopicOrPartition, "unknown topic #{topic}"
         end
+        Array.new(partitions.count) { topic }.zip(partitions)
+      end
+
+      partitions_per_member = [1, (topic_partitions.count / members.count)].max
+      partitions_for_each_member = topic_partitions.sort.each_slice(partitions_per_member).to_a
 
-        partitions_per_member = partitions.group_by {|partition_id|
-          partition_id % members.count
-        }.values
+      if partitions_for_each_member.count > members.count
```

Comment:
I don't get the idea behind this code. Isn't it a little bit complicated? The problem with old code is that it is topic-scoped. So, to make the partitions spread equally, we just need to split by all topics, not single one. Something like this:

```ruby
topic_partitions = topics.flat_map do |topic|
  begin
    @cluster.partitions_for(topic).map { |p| [topic, p.partition_id }
  rescue UnknownTopicOrPartition
    raise UnknownTopicOrPartition, "unknown topic #{topic}"
  end
end

partitions_per_member = partitions.group_by.each_index do |_, index| 
  index % members.count 
end.values

members.zip(partitions_per_member).each do |member_id, member_partitions|
  unless member_partitions.nil?
    member_partitions.each do |topic, partition|
      group_assignment[member_id].assign(topic, [partition])
    end
  end
end
```

Reply:
You're 100% right. I thought I had a simpler implementation but then it missed a whole load of edge cases and fixing them made it all vastly less simple. Updated.